component/org.wso2.carbon.identity.oauth2.grant.token.exchange/src/main/java/org/wso2/carbon/identity/oauth2/grant/token/exchange/TokenExchangeGrantHandler.java

```diff
@@ -136,6 +155,219 @@ public boolean validateGrant(OAuthTokenReqMessageContext tokReqMsgCtx) throws Id
+            throws IdentityOAuth2Exception {
+
+        // Retrieve the signed JWT object from the request parameters
+        SignedJWT signedJWT = getSignedJWT(requestParams.get(TokenExchangeConstants.SUBJECT_TOKEN));
+        if (signedJWT == null) {
+            // If no valid subject token found, handle the exception
+            handleException(OAuth2ErrorCodes.INVALID_REQUEST,
+                    "No Valid subject token was found for " + TokenExchangeConstants.TOKEN_EXCHANGE_GRANT_TYPE);
+        }
+
+        // Extract claims from the JWT
+        JWTClaimsSet claimsSet = getClaimSet(signedJWT);
+        if (claimsSet == null) {
+            // If claim values are empty, handle the exception
+            handleException(OAuth2ErrorCodes.INVALID_REQUEST, "Claim values are empty in the given JSON Web Token");
+        }
+
+        // Validate mandatory claims
+        String subject = resolveSubject(claimsSet);
+        validateMandatoryClaims(claimsSet, subject);
+
+        impersonator = resolveImpersonator(claimsSet);
+        if (StringUtils.isBlank(impersonator)) {
+            // If claim values are empty, handle the exception
+            handleException(OAuth2ErrorCodes.INVALID_REQUEST, "Subject of actor token and impersonator are different.");
+        }
+
+        // Validate if the subject token is signed by the AS
+        if (!validateTokenSignature(signedJWT, tenantDomain)) {
+            handleException(OAuth2ErrorCodes.INVALID_REQUEST, "Signature or Message Authentication invalid");
+        }
+
+        checkJWTValidity(claimsSet);
+
+        // Validate the audience of the subject token
+        List<String> audiences = claimsSet.getAudience();
+        if (!validateSubjectTokenAudience(audiences, tokReqMsgCtx)) {
+            handleException(TokenExchangeConstants.INVALID_TARGET, "Invalid audience values provided");
+        }
+
+        // Validate the issuer of the subject token
+        String jwtIssuer = claimsSet.getIssuer();
+        validateTokenIssuer(jwtIssuer, tenantDomain);
+
+        tokReqMsgCtx.addProperty(IMPERSONATING_ACTOR, impersonator);
+        tokReqMsgCtx.setScope(getScopes(claimsSet, tokReqMsgCtx));
+    }
+
+    /**
+     * Retrieves the scopes claim from the JWTClaimsSet object and splits it into an array of individual scope strings.
+     * Assumes that the scopes claim is represented as a space-delimited string.
+     *
+     * @param claimsSet    The JWTClaimsSet object containing the token claims.
+     * @param tokReqMsgCtx
+     * @return An array of individual scope strings extracted from the scopes claim.
+     */
+    private String[] getScopes(JWTClaimsSet claimsSet, OAuthTokenReqMessageContext tokReqMsgCtx) {
+
+        String[] requestedScopes = tokReqMsgCtx.getScope();
+        String[] approvedScopes = ((String) claimsSet.getClaims().get(SCOPE)).split("\\s+");
+        if (requestedScopes == null) {
+            return approvedScopes;
+        } else {
```

Comment:
else is not needed.
 String[] requestedScopes = tokReqMsgCtx.getScope();
        String[] approvedScopes = ((String) claimsSet.getClaims().get(SCOPE)).split("\\s+");
        if (requestedScopes == null) {
            return approvedScopes;
        } 
return filterRequestedScopes(requestedScopes, approvedScopes);
}

Reply:
+1 ref: https://github.com/wso2-extensions/identity-oauth2-grant-token-exchange/pull/26/commits/d1762473ae0c89124fb4cd7a6158649e542a595f